spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/lifecycle/MemoizedValueAdapter.kt

```diff
@@ -33,51 +34,52 @@ sealed class MemoizedValueAdapter<T>(
         }
     }
 
-    class GroupCachingModeAdapter<T>(
-        factory: () -> T,
-        destructor: (T) -> Unit
-    ) : MemoizedValueAdapter<T>(factory, destructor) {
+    internal abstract fun registerEagerInitializer(root: Root)
+}
 
-        private val stack = mutableListOf<Cached<T>>()
+class GroupCachingModeAdapter<T>(
+    factory: () -> T,
+    destructor: (T) -> Unit
+) : MemoizedValueAdapter<T>(factory, destructor) {
 
-        override fun beforeExecuteGroup(group: GroupScope) {
-            stack.add(0, cached)
-            cached = Cached.Empty
-        }
+    private val stack = mutableListOf<Cached<T>>()
 
-        override fun afterExecuteGroup(group: GroupScope) {
-            val cached = this.cached
-            if (cached is Cached.Value<T>) {
-                destructor(cached.value)
-            }
-            if (stack.isNotEmpty()) {
-                this.cached = stack.removeAt(0)
-            }
+    override fun registerEagerInitializer(root: Root) {
+        root.beforeGroup {
```

Comment:
Is this due to the runtime pre-pending lifecycle listeners? https://github.com/spekframework/spek/blob/2.x/spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/lifecycle/LifecycleManager.kt#L17

Reply:
Yes, partly. As I explained in the previous PR, there were two ordering issues, this seems to be the easiest way to solve both of those them, to keep the order of `memoized` initialization and fixtures consistent with their declaration order in tests.

Reply:
I'll get back to you why I did it(prepend) that way. I'm thinking if we could just change `memoized` to use the fixtures or change the implementation of fixtures so that you can weave them in between `memoized` calls and get the _expected_ execution sequence.

Reply:
found it https://github.com/spekframework/spek/pull/183/files 🤦‍♂️ , what was I thinking.. Anyway we have to ways to fix this:

1. Change `memoized` implementation to use fixtures.
2. Change fixtures implementation to register a LifecycleListener for each call instead of having one LifecycleListener (`FixturesAdapater`)- and stop prepending listeners.

If we can make sure that the execution order of `memoized` and `fixtures` are based on declartion order then we can stop prepending listeners. I'm leaning more to option 1, wdyt?

Reply:
I apologize for the delayed response. Both options works fine to me, though implementing `memoized` using fixtures seems the easiest, and is the least likely to break backwards compatibility. I need to mention that memoized still registers its own `LifecycleListener`s for clearing its values and invoking the destructors. I think we should either use fixtures for invoking the destructors, or keep them as listeners, and do as you described in `2.`.

Reply:
@raniejade I'm sorry to tag you, but you haven't responded in 4 days now.

Reply:
@feature sorry for the delayed response! I'm leaning towards using fixtures all the way, so instead of `memoized` registering `LifecycleListerner`s it just register fixtures (depending on the `CachingMode`).

Reply:
@feature just realized that we cant use fixtures with `CachingMode.GROUP`. `beforeGroup` is only executed once not per group.

Reply:
Hmm. Do you got any other ideas? I have been rather busy with life, but I'll take another deeper look in the issue when I have time.

Reply:
Trying to align fixtures and memoized, eventually adding a `beforeEachGroup` fixture. I think that should help us implement this.